tests/unit/misc/test_mhtml.py

```diff
@@ -0,0 +1,268 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+import io
+import textwrap
+import re
+
+import pytest
+
+from qutebrowser.misc import mhtml
+
+@pytest.fixture(autouse=True)
+def patch_uuid(monkeypatch):
+    monkeypatch.setattr("uuid.uuid4", lambda: "UUID")
+
+
+class Checker:
+
+    """A helper to check mhtml output.
+
+    Attrs:
+        fp: A BytesIO object for passing to MHTMLWriter.write_to.
+    """
+
+    def __init__(self):
+        self.fp = io.BytesIO()
+
+    @property
+    def value(self):
+        return self.fp.getvalue()
+
+    def expect(self, expected):
+        actual = self.value.decode("ascii")
+        # Make sure there are no stray \r or \n
+        assert re.search(r"\r[^\n]", actual) is None
+        assert re.search(r"[^\r]\n", actual) is None
+        actual = actual.replace("\r\n", "\n")
+        expected = textwrap.dedent(expected).lstrip("\n")
+        assert expected == actual
+
+
+@pytest.fixture
+def checker():
+    return Checker()
+
+
+def test_quoted_printable_umlauts(checker):
+    content = "Die süße Hündin läuft in die Höhle des Bären"
+    content = content.encode("iso-8859-1")
+    writer = mhtml.MHTMLWriter(root_content=content,
+                               content_location="localhost",
+                               content_type="text/plain")
+    writer.write_to(checker.fp)
+    checker.expect("""
+    Content-Type: multipart/related; boundary="---=_qute-UUID"
+    MIME-Version: 1.0
+
+    -----=_qute-UUID
+    Content-Location: localhost
+    MIME-Version: 1.0
+    Content-Type: text/plain
+    Content-Transfer-Encoding: quoted-printable
+
+    Die=20s=FC=DFe=20H=FCndin=20l=E4uft=20in=20die=20H=F6hle=20des=20B=E4ren
+    -----=_qute-UUID--
+    """)
+
+
+@pytest.mark.parametrize("header, value", [
+    ("content_location", "http://brötli.com"),
+    ("content_type", "text/pläin"),
+])
+def test_refuses_non_ascii_header_value(checker, header, value):
+    defaults = {
+        "root_content": b"",
+        "content_location": "http://example.com",
+        "content_type": "text/plain",
+    }
+    defaults[header] = value
+    writer = mhtml.MHTMLWriter(**defaults)
+    with pytest.raises(UnicodeEncodeError) as e:
+        writer.write_to(checker.fp)
+        assert "'ascii' codec can't encode" in str(e)
```

Comment:
I'm not sure what `str()` on an `ExceptionInfo` instance will do exactly (and it's [undocumented](http://pylib.readthedocs.org/en/latest/code.html#py-code-exceptioninfo)) - please do something like this instead:

```python
with pytest.raises(UnicodeEncodeError) as excinfo:
    writer.write_to(checker.fp)
    assert "'ascii' codec can't encode" in str(excinfo.value)
```

(I always call it `excinfo` as it's not the exception itself)